academy/ibc-dev/ibcapp_steps.md

```diff
@@ -0,0 +1,671 @@
+
+</HighlightBox>
+
+#### Packet callbacks
+
+The general application packet flow was discussed in [a previous section](https://tutorials.cosmos.network/academy/4-ibc/channels.html#application-packet-flow). As a refresher, let's take a look at the diagram: 
+
+![packet flow](./images/packetflow.png)
+
+The packet callbacks in the packet flow can now be identified by investigating the `IBCModule` interface.
+
+##### Sending packets
+
+Modules **do not send packets through callbacks**, since modules initiate the action of sending packets to the IBC module, as opposed to other parts of the packet flow where messages sent to the IBC module must trigger execution on the port-bound module through the use of callbacks. Thus, to send a packet a module simply needs to call `SendPacket` on the `IBCChannelKeeper`.
+
+<HighlightBox type="warning">
+
+In order to prevent modules from sending packets on channels they do not own, IBC expects modules to pass in the correct channel capability for the packet's source channel.
+
+</HighlightBox>
+
+<HighlightBox type="reading">
+
+For advanced readers, more on capabilities can be found in the [ibc-go docs](https://ibc.cosmos.network/main/ibc/overview.html#capabilities) or in the [ADR on the Dynamic Capability Store](https://github.com/cosmos/cosmos-sdk/blob/6aaf83c894e917836a047b0399dd70a95fd2710d/docs/architecture/adr-003-dynamic-capability-store.md).
+
+</HighlightBox>
+
+##### Receiving packets
+
+To handle receiving packets, the module must implement the `OnRecvPacket` callback. This gets invoked by the IBC module after the packet has been proved valid and correctly processed by the IBC keepers. Thus, the `OnRecvPacket` callback only needs to worry about making the appropriate state changes given the packet data without worrying about whether the packet is valid or not.
+
+Modules may return to the IBC handler an acknowledgement which implements the `Acknowledgement` interface. The IBC handler will then commit this acknowledgement of the packet so that a relayer may relay the acknowledgement back to the sender module.
+
+The state changes that occurred during this callback will only be written if:
+
+- The acknowledgement was successful as indicated by the `Success()` function of the acknowledgement.
+- The acknowledgement returned is nil, indicating that an asynchronous process is occurring.
+
+<HighlightBox type="note">
+
+Applications which process asynchronous acknowledgements must handle reverting state changes when appropriate. Any state changes that occurred during the `OnRecvPacket` callback will be written for asynchronous acknowledgements.
+
+</HighlightBox>
+
+In `x/leaderboard/module_ibc.go` scaffolded by Ignite CLI we find `OnRecvPacket`:
+
+```go
+// OnRecvPacket implements the IBCModule interface
+func (am AppModule) OnRecvPacket(
```

Comment:
Maybe just copy again the full implementation of the callback here?

Reply:
I'm a bit concerned about total length if we do. Could use an ExpansionPanel as I've started to do (but the code is already present in full so not sure if it's required to have it in full once more)